File: frontend/src/types/task.ts
export type TaskStatus = 'todo' | 'in_progress' | 'review' | 'completed' | 'on_hold' | 'cancelled'
export type TaskPriority = 'low' | 'medium' | 'high' | 'urgent'

export interface Task {
  id: string
  title: string
  description?: string
  assigned_to: string
  assigned_by: string
  allocated_hours: number
  actual_hours: number
  priority: TaskPriority
  status: TaskStatus
  progress_percentage: number
  start_date?: string
  due_date?: string
  completed_at?: string
  is_overdue: boolean
  task_category?: string
  tags?: string[]
  created_at: string
  updated_at: string
  // Joined fields
  assigned_to_name?: string
  assigned_by_name?: string
  efficiency_percentage?: number
}

export interface TaskCreate {
  title: string
  description?: string
  assigned_to: string
  allocated_hours: number
  priority: TaskPriority
  start_date?: string
  due_date?: string
  task_category?: string
  tags?: string[]
}

export interface TaskUpdate {
  title?: string
  description?: string
  status?: TaskStatus
  progress_percentage?: number
  priority?: TaskPriority
  allocated_hours?: number
  start_date?: string
  due_date?: string
  task_category?: string
  tags?: string[]
}

export interface TaskStats {
  total_tasks: number
  completed_tasks: number
  pending_tasks: number
  overdue_tasks: number
  completion_rate: number
  avg_efficiency: number
  on_time_rate: number
}

File: backend/app/schemas/task.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime, date
from enum import Enum

class TaskStatus(str, Enum):
    TODO = "todo"
    IN_PROGRESS = "in_progress"
    REVIEW = "review"
    COMPLETED = "completed"
    ON_HOLD = "on_hold"
    CANCELLED = "cancelled"

class TaskPriority(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"

class TaskBase(BaseModel):
    title: str = Field(..., max_length=255)
    description: Optional[str] = None
    assigned_to: str
    allocated_hours: float = Field(..., gt=0)
    priority: TaskPriority = TaskPriority.MEDIUM
    start_date: Optional[date] = None
    due_date: Optional[date] = None
    task_category: Optional[str] = None
    tags: Optional[List[str]] = None

class TaskCreate(TaskBase):
    pass

class TaskUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    status: Optional[TaskStatus] = None
    progress_percentage: Optional[int] = Field(None, ge=0, le=100)
    priority: Optional[TaskPriority] = None
    start_date: Optional[date] = None
    due_date: Optional[date] = None
    allocated_hours: Optional[float] = Field(None, gt=0)
    task_category: Optional[str] = None
    tags: Optional[List[str]] = None

class TaskStatusUpdate(BaseModel):
    status: TaskStatus
    progress_percentage: Optional[int] = Field(None, ge=0, le=100)

class TaskResponse(TaskBase):
    id: str
    assigned_by: str
    actual_hours: float = 0
    status: TaskStatus
    progress_percentage: int = 0
    completed_at: Optional[datetime] = None
    is_overdue: bool = False
    created_at: datetime
    updated_at: datetime
    # Present on TaskResponse (not just TaskDetailResponse) because the
    # list endpoint (`GET /tasks/`, used by the Admin Tasks table) returns
    # TaskResponse, and that table needs the assignee's name per row.
    assigned_to_name: Optional[str] = None
    
    class Config:
        from_attributes = True

class TaskDetailResponse(TaskResponse):
    assigned_by_name: Optional[str] = None
    efficiency_percentage: Optional[float] = None
    timeliness_status: Optional[str] = None
    time_logs: Optional[List] = []

class TaskStatsResponse(BaseModel):
    total_tasks: int
    completed_tasks: int
    pending_tasks: int
    overdue_tasks: int
    completion_rate: float
    avg_efficiency: float
    on_time_rate: float